### PR TITLE
[perf] measure paged attention token prefill/generate time

### PR DESCRIPTION
**Pull request by xy12181**
```
$ ./MaxText/tests/inference/test_llama2_7b_int8.sh
```
```
prefill: stream=0, #tokens=4/16, time=7906.16 ms
generate: token=#17, time=8804.44 ms
generate: token=#18, time=2.51 ms
generate: token=#19, time=1.99 ms
generate: token=#20, time=1.74 ms
generate: token=#21, time=1.99 ms
generate: token=#22, time=2.20 ms
generate: token=#23, time=1.74 ms
generate: token=#24, time=1.74 ms
generate: token=#25, time=1.77 ms
generate: token=#26, time=3.11 ms
generate: token=#27, time=11.25 ms
generate: token=#28, time=11.09 ms
generate: token=#29, time=11.12 ms
generate: token=#30, time=11.07 ms
generate: token=#31, time=11.08 ms
generate: token=#32, time=11.13 ms
```

```
$ ./MaxText/tests/inference/test_llama2_7b_bf16.sh
```
```
prefill: stream=0, #tokens=4/16, time=3002.58 ms
generate: token=#17, time=3368.66 ms
generate: token=#18, time=1021.62 ms
generate: token=#19, time=1.64 ms
generate: token=#20, time=1.67 ms
generate: token=#21, time=1.37 ms
generate: token=#22, time=1.76 ms
generate: token=#23, time=1.42 ms
generate: token=#24, time=1.43 ms
generate: token=#25, time=1.37 ms
generate: token=#26, time=1.41 ms
generate: token=#27, time=4.33 ms
generate: token=#28, time=9.34 ms
generate: token=#29, time=9.35 ms
generate: token=#30, time=9.27 ms
generate: token=#31, time=9.29 ms
generate: token=#32, time=9.32 ms

```

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by xy12181** [comment]
@vipannalla  thank you for merging my previous PR. here goes another one. :)

File: MaxText/decode.py [hunk 1]
```diff
@@ -64,9 +69,13 @@ def main(argv: Sequence[str]) -> None:
...
   sampled_tokens_list = []
   sampled_tokens_list.append(first_token)
-  for _ in steps:
+  for i in steps:
     rng, rng_generate = jax.random.split(rng)
+    gt0 = perf_counter()
     decode_state, sampled_tokens = engine.generate(params, decode_state, rng=rng_generate)
+    gt1 = perf_counter()
+    msec = (gt1 - gt0) * 1000.0
+    print(f"generate: token=#{i+1}, time={msec:.2f} ms")
```
Comment (wyzhang) [inline MaxText/decode.py:78]: nit: token->step

Might be more clear to use the term `step` rather than `token` here

**Review by wyzhang — APPROVED** [review]

**Review by vipannalla — APPROVED** [review]

File: MaxText/decode.py (hunk 1 above)
Comment (xy12181) [reply MaxText/decode.py:78]: ack

Comment (wyzhang) [reply MaxText/decode.py:78]: [Minor] Not sure if I follow. Just to clarify I was suggesting to change the message to the following. This is nitpicking, what you have is fine.

print(f"generate: step={i}, time={msec:.2f} ms")

Comment (xy12181) [reply MaxText/decode.py:78]: log line#    step#(i)      token# (i+1)   // i+1 is slightly better than i (if we use i, the 17th token will print 16)
prefill           1                   1                 1...16 (4 real + 12 padding)
generate     2                  16                17
generate     3                  17                18
generate     4                  18                19

you can see the step# (i) is actually the (i+1)th token generated. hence i used token. by the meaning of word step, one natural guess is the log line sequence number: the first step (prefill), the second step (generate), etc.

Comment (wyzhang) [reply MaxText/decode.py:78]: SG

**Comment by xy12181** [comment]
@vipannalla   could you please trigger merge of this PR? thanks!

**Review by vipannalla — APPROVED** [review]
looks good

**Comment by xy12181** [comment]
@vipannalla  i rebased this PR to the latest main branch (#1403) so it is easy for you to merge.

**Review by vipannalla — APPROVED** [review]
